[driver] Implement a Windows toolchain for cross-compiling

troughton opened the pull request:
In this PR, I've implemented an initial Windows toolchain that has been tested to work in cross-compiling from Linux to Windows. 

When combined with https://github.com/apple/swift/pull/13140, this enables an invocation like:

```
swift -target x86_64-unknown-windows-msvc -use-ld=lld main.swift -o SwiftTest.exe
```

to produce a Windows executable (note that the executable itself will still run into separate issues).

I've followed the pattern of the `GenericUnix` toolchain here, which *does* mean there's currently a fair amount of code duplication. I don't know how much will need to change for native builds, and nor do I know how many of the extra arguments (such as ASAN or producing a static executable) will work as-is. However, I thought that Windows is sufficiently different enough from the existing OSes to justify its own toolchain and expect it to diverge further as it is expanded, making this a good starting point.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1104,6 +1113,16 @@ getSanitizerRuntimeLibNameForDarwin(StringRef Sanitizer,
+  return (Twine("clang_rt.")
+      + Sanitizer + "-"
+      + Triple.getArchName()
+      + (shared ? ".dll" : ".lib")).str();
```
I'm not sure that this is right.  You never link against the `.dll` but rather the `.lib`.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1159,6 +1187,20 @@ addLinkRuntimeLibForDarwin(const ArgList &Args, ArgStringList &Arguments,
+  getRuntimeLibraryPath(Dir, Args, TC);
+  // Remove platform name.
+  llvm::sys::path::remove_filename(Dir);
+  llvm::sys::path::append(Dir, "clang", "lib", "windows");
```
Isn't this effectively `getClangLibraryPathOnWindows`?

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1191,6 +1233,23 @@ addLinkSanitizerLibArgsForDarwin(const ArgList &Args,
+  Arguments.push_back("-lc++");
+  // Add explicit dependency on -lc++abi, as -lc++ doesn't re-export
+  // all RTTI-related symbols that are used.
+  Arguments.push_back("-lc++abi");
```
Why the assumption on the library names here and the spelling for the argument?  Furthermore, Windows doesn't use `libc++abi` ever but rather vcruntime.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+
+  addPathEnvironmentVariableIfNeeded(II.ExtraEnvironment, "PATH",
+                                     ":", options::OPT_L, context.Args,
+                                     runtimeLibraryPath);                                            
```
Please don't do this, this way (via the environment).  We use use `-libpath` or `-L`.

jrose-apple replied:
That doesn't work for the interpreter.

troughton replied:
Fixing this is probably out of scope for the PR as I'm only able to test cross-compiling at the moment, so testing the interpreter isn't an option.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  Arguments.push_back(
+      context.Args.MakeArgString(context.Output.getPrimaryOutputFilename()));
+
+  return {"swift-autolink-extract", Arguments};
```
This feels wrong.  We shouldn't be using `swift-autolink-extract` on Windows at all.  We have auto-linking support, and we should make that work instead.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+      Clang = context.Args.MakeArgString(toolchainClang.get());
+    }
+
+    // Look for binutils in the toolchain folder.
```
We shouldn't need binutils at all really.  Everything is possible to do via the combination of LLVM/clang/lld or clang/Visual Studio.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  // Add the runtime library link path, which is platform-specific and found
+  // relative to the compiler.
+  if (!(staticExecutable || staticStdlib)) {
+    // FIXME: We probably shouldn't be adding a libpath here unless we know
```
I don't understand this comment.  What do you mean that it won't be copied?  This is used during linking, not during the runtime.  Furthermore, I believe that clang in the GNU driver mode will transliterate the `-L` appropriately, so no need for the `-Xlinker /LIBPATH:...`.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  // Pass them through to the linker here for convenience.
+  const char *VCToolsInstallDir = ::getenv("VCToolsInstallDir");
+  const char *UniversalCRTSdkDir = ::getenv("UniversalCRTSdkDir");
+  const char *UCRTVersion = ::getenv("UCRTVersion");
```
Ugh, this is rather annoying to replicate here, especially since the variables and the paths are subject to Microsoft's control.  We should just request that the user pass in the appropriate paths.

troughton replied:
The rationale here is that if the user has cross-compiled Swift they'll already have these environment variables set up; it's a nuisance to have to manually copy across `LIB` and `INCLUDE` from an MSVC shell. This eases a significant burden on the user – is it really that bad to have it in another place in the Swift project, at least while cross-compilation is the most reliable Windows toolchain?

troughton replied:
Actually, thinking about it more, I think you're right here. I've been doing Swift builds regularly so always have those variables set, but someone using a build will just want to have e.g. `INCLUDE` and `LIB` set in their shell configuration file, and it's easy enough to provide instructions on how to do that.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  if (!VCToolsDir.empty()) {
+    Arguments.push_back("-Xlinker");
+    Arguments.push_back(context.Args.MakeArgString("/LIBPATH:" + llvm::Twine(VCToolsInstallDir) 
+        + "/Lib/" + tripleWinArchName));
```
See previous point of `-L`.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+          + "/Lib/" + llvm::Twine(UCRTVer) + "/ucrt/" + tripleWinArchName));
+      Arguments.push_back("-Xlinker");
+      Arguments.push_back(context.Args.MakeArgString("/LIBPATH:" + llvm::Twine(UCRTDir) 
+          + "/Lib/" + llvm::Twine(UCRTVer) + "/um/" + tripleWinArchName));
```
And again

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    Arguments.push_back(context.Args.MakeArgString(context.OI.SDKPath));
+  }
+
+  // Add any autolinking scripts to the arguments
```
See previous point about auto-linking.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  }
+  else {
+    Arguments.push_back(context.Args.MakeArgString(SharedRuntimeLibPath));
+    Arguments.push_back("-lswiftCore");
```
This should be the absolute path to the `swiftCore.lib` as that is in a well defined location I believe.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    Arguments.push_back("-lswiftCore");
+  }
+  
+  // Explicitly pass the target to the linker
```
This should be to the driver, the linker does not read the target.

compnerd reviewed (changes requested):
I think that we should split out the implementation into a new `.cpp`.

troughton commented:
@compnerd Okay, I'll start making those adjustments. Just a note that whenever there's a 'why have you done this' it's because I started with the GenericUnix toolchain and just tweaked until it worked; this was just a starting point and I was fulling expecting further changes.

troughton commented:
@jrose-apple Would you want this split out into a separate CPP file, and, if so, should the GenericUnix and Darwin toolchains also have their own CPP files?

troughton commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,223 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  for (const Arg *arg : context.Args.filtered(options::OPT_F,
+                                              options::OPT_Fsystem)) {
+    if (arg->getOption().matches(options::OPT_Fsystem))
+      Arguments.push_back("-iframework");
```
Can we ever get frameworks being passed to the linker in a Windows context?

compnerd replied:
The only thing that is interesting about a framework is the fact that they have a particular layout.  I don't see a strong reason to not forward those along to clang and have clang just transliterate them into the appropriate `-I` and `-L` options.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1191,6 +1225,20 @@ addLinkSanitizerLibArgsForDarwin(const ArgList &Args,
+                                 ArgStringList &Arguments,
+                                 StringRef Sanitizer,
+                                 const ToolChain &TC,
+                                 bool shared = true
```
Nitpick: misaligned arguments here, and the close paren should follow the last argument instead of going on the next line.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+ToolChain::InvocationInfo
+toolchains::Windows::constructInvocation(const InterpretJobAction &job,
+                                             const JobContext &context) const {
+  InvocationInfo II = ToolChain::constructInvocation(job, context);
```
No need for this now that you're inheriting directly from ToolChain.

jrose-apple replied:
I see this is vestigial because @compnerd had you remove the LD_LIBRARY_PATH trick. Unfortunately we'll need to come up with something similar for Windows if we want the interpreter to work with -L. We're probably not doing the right thing on Windows anyway, though (`dlopen`).

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+
+  // Configure the toolchain.
+  // By default, use the system clang++ to link.
+  const char * Clang = "clang++";
```
Has there been any discussion about using `cl` directly? Do we get any benefit from using Clang?

troughton replied:
I think there are tradeoffs to each approach. With `cl`, the user doesn't need Clang installed, but Clang does provide things like ASAN that MSVC is missing. If we also might end up with a dependency on Compiler-RT (see https://github.com/apple/swift/pull/13234) then it'd probably be worth sticking with Clang.

As for using the `clang-cl` vs. `clang` driver: I don't have any strong opinions there. I suppose there'd be some benefit in consistency with other platforms, but if the consensus is that `clang-cl` is preferred I can easily switch over the commands.

troughton replied:
I suppose one option would be to use `clang-cl` by default and fallback to `cl` if Clang isn't installed.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+
+  StringRef tripleWinArchName;
+  switch (getTriple().getArch()) {
+    case llvm::Triple::x86:
```
Nitpick: `case` labels are not indented in LLVM style.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  StringRef UCRTVer(UCRTVersion);
+
+  if (!UCRTDir.empty() && !UCRTVer.empty()) {
+      Arguments.push_back("-L");
```
Nitpick: accidental double-indent here.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    Arguments.push_back(context.Args.MakeArgString(StaticRuntimeLibPath));
+
+    SmallString<128> linkFilePath = StaticRuntimeLibPath;
+    llvm::sys::path::append(linkFilePath, "static-executable-args.lnk");
```
Does this work on Windows?

troughton replied:
The static stdlib isn't currently building on Windows, but thinking about it more I think we'd end up passing the same arguments here whether we're building static or shared, so this section should be removed. @compnerd, thoughts?

jrose-apple replied:
I meant linker scripts in general, but that's also a reasonable question.

compnerd replied:
Yeah, I think I made a similar point when I looked through this, that the invocation should be similar.

@jrose-apple linker scripts are not supported on Windows, but this isn't a linker script, it is a response file, which is supported (link and lld both support it, and those are the only viable linkers atm).

jrose-apple replied:
Whoops, of course. Thanks, Saleem.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  }
+  
+  // Explicitly pass the target to the driver
+  Arguments.push_back(context.Args.MakeArgString("--target=" + getTriple().str()));
```
You already have this above.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1519,222 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    Arguments.push_back(context.Args.MakeArgString(LibProfile));
+    Arguments.push_back(context.Args.MakeArgString(
+        Twine("-u", llvm::getInstrProfRuntimeHookVarName())));
+  }
```
Do all of these things work on Windows?

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1690,7 +1954,7 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   }
   
-  // Explicitly pass the target to the linker
+  // Explicitly pass the target to the driver
```
"the driver" means "the Swift driver" in this context. If you don't like "the linker", please just say "Clang" here.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1191,6 +1225,20 @@ addLinkSanitizerLibArgsForDarwin(const ArgList &Args,
+                                 bool shared = true
+                                 ) {
+  // Sanitizer runtime libraries requires C++.
+  Arguments.push_back("-lc++");
```
Doesn't everything need C++, not just the sanitizers?

troughton replied:
This is copied across from the Darwin implementation. I haven't had a chance (yet) to test the sanitisers, but everything else works fine without the `-lc++` flag – it might not even be necessary with the sanitisers.

In general, do we want to err on the side of 'untested but might work', or 'only present if it's guaranteed to work,' considering there's a lot on Windows that won't be able to be tested properly for a while?

jrose-apple replied:
My thoughts are that no one will ever remove something unnecessary, but they will add what's necessary. Therefore I'd rather err on the minimal side.

jrose-apple reviewed (commented):
It probably does make sense to split up ToolChains.cpp at this point. That should happen in a no-functionality-change commit, though, and that can be after this PR.

troughton commented:
I've just run `clang-format` and applied the changes to the Windows sections of the file, so some comments above might be marked as outdated when they're still relevant. I should note that there's a lot of inconsistencies in this file in terms of formatting (of which my changes inherited a few), so it might be worth running `clang-format` through it all in the process of splitting it up.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1104,6 +1113,12 @@ getSanitizerRuntimeLibNameForDarwin(StringRef Sanitizer,
 }
 
+static std::string getSanitizerRuntimeLibNameForWindows(
+    StringRef Sanitizer, const llvm::Triple &Triple, bool shared = true) {
```
Shared is an unused parameter.

compnerd commented on lib/Driver/Driver.cpp:
```diff
@@ -221,8 +221,12 @@ makeToolChain(Driver &driver, const llvm::Triple &target) {
+      return llvm::make_unique<toolchains::Cygwin>(driver, target);
+    } else {
+      return llvm::make_unique<toolchains::Windows>(driver, target);
+    }
```
Unnecessary braces.

troughton replied:
This matches the `target.isAndroid` conditional above. What's the Swift project style on this? Should I change both of them to remove the braces?

jrose-apple replied:
We're based on LLVM style, which says no braces for single-line if-statements. We swing pretty heavily on the side of having them whenever it's a grey area, though:

- If you have comments inside the body, provide the braces.
- If the line wraps, people usually provide the braces.
- If another branch of the `if` has braces, people usually add braces to all of them.
- If the *condition* wraps, sometimes people provide braces.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1191,6 +1222,20 @@ addLinkSanitizerLibArgsForDarwin(const ArgList &Args,
+                                              const ToolChain &TC,
+                                              bool shared = true) {
+  // Sanitizer runtime libraries requires C++.
+  Arguments.push_back("-lc++");
```
This naming is still wrong.  Plus, what if Im using `msvcprt` for the C++ runtime?

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1516,220 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    if (auto toolchainClang =
+            llvm::sys::findProgramByName("clang++", {toolchainPath})) {
+      Clang = context.Args.MakeArgString(toolchainClang.get());
+    }
```
If we use `clang++`, we shouldn't be adding the C++ runtime link.

compnerd replied:
I think that we should always do a `findProgramByName` for clang++.  That way when you do `-v` you can see which clang is being used.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1516,220 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  } else if (context.Args.hasFlag(options::OPT_static_stdlib,
+                                  options::OPT_no_static_stdlib, false)) {
+    staticStdlib = true;
+  }
```
Initialize the variables at declaration time rather than the separate conditional?

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1571,16 +1567,10 @@ toolchains::Windows::constructInvocation(const LinkJobAction &job,
+  bool staticExecutable = context.Args.hasFlag(
+      options::OPT_static_executable, options::OPT_no_static_executable, false);
+  bool staticStdlib = context.Args.hasFlag(
+      options::OPT_static_stdlib, options::OPT_no_static_stdlib, false);
```
This isn't exactly equivalent: previously `staticExecutable` implied `!staticStdlib`. But that doesn't actually make sense to me; shouldn't it be the other way around?

troughton commented:
I was playing around with the different options and seeing if I could test the different `-static-executable` and `-static-stdlib` flags and came to the conclusion that it's probably best to leave both out of this PR. The static standard library isn't building correctly yet on Windows and I'm not confident enough in my understanding on how linking works on Windows to justify keeping that code path in.

With regards to the `.lnk` files, I believe they're not getting produced correctly on anything other than Linux at the moment, but when they *are* there for Windows the code should be able to be copied pretty much verbatim from the `GenericUnix` toolchain.

I believe I've addressed the other concerns. Whether we should be using `cl` or `clang++` is an open discussion point, but for now, since `clang++` is the only option, I'm leaving in the options for the various sanitisers and for optimisation profile generation. I see no reason why they wouldn't work, although they're again hard to test in a cross-compile environment.

compnerd commented:
Can you please add some tests to ensure that the various codepaths generate the proper tool invocations?

troughton commented:
Sure. I've briefly looked through test/Driver and there's quite a lot in there – this is the first functionality I've added to Swift so I'm not familiar with the all the testing commands and output check tools. Is there anywhere in particular you could point me to for other platforms that would be a good starting point?

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1471,6 +1512,124 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+                                         const JobContext &context) const {
+  InvocationInfo II = ToolChain::constructInvocation(job, context);
+
+  return II;
```
Let's just take this out if we're not going to be doing anything with it right now.

jrose-apple reviewed (commented):
Some tests to add to / steal from:
- Driver/linker.swift has a bunch of simple link lines just to test that we get to linking at all
- Driver/linker-autolink-extract.swift probably works unmodified with a Windows triple
- Driver/sdk.swift could be relevant
- Driver/sanitizers.swift is probably worth adding to, since you included sanitizer logic here

troughton commented:
I've added tests in those files and fixed a couple of other small issues I've found. All of the tests pass locally. Is there anywhere in particular I've missed coverage?

troughton commented:
Checking back on this PR - are we okay to merge this?

jrose-apple reviewed (commented):
Looks good from my end. Saleem?

troughton commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -139,7 +139,7 @@ OptionSet<SanitizerKind> swift::parseSanitizerArgValues(
 
   // Sanitizers are only supported on Linux or Darwin.
-  if (!(Triple.isOSDarwin() || Triple.isOSLinux())) {
+  if (!(Triple.isOSDarwin() || Triple.isOSLinux() || Triple.isOSWindows())) {
```
I've changed this to "Check that we're one of the known supported targets for sanitizers."

troughton commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1025,6 +1025,15 @@ static void getClangLibraryPathOnDarwin(SmallVectorImpl<char> &libPath,
   llvm::sys::path::append(libPath, "clang", "lib", "darwin");
 }
 
+static void getClangLibraryPathOnWindows(SmallVectorImpl<char> &libPath,
```
I'd be happy to take a look at splitting up this file and refactoring some of the common functionality in an NFC follow-up PR – I think there's some fairly straightforward things that would help.

troughton commented:
I've updated this PR to include static linking support, since I now have that working locally on Windows. I also added `swiftrt.o` to the linker invocation since that's now required for COFF binaries to correctly initialise.

troughton commented:
@compnerd Does this look okay to you now?

troughton commented:
@jrose-apple If @compnerd doesn't have any complaints, are we okay to merge this? There've been a couple of small changes (static linking support and adding the COFF registration object) since you likely last looked at it.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1493,6 +1534,128 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    getRuntimeStaticLibraryPath(StaticRuntimeLibPath, context.Args, *this);
+    
+    // Since Windows has separate libraries per architecture, link against the
+    // architecture specific version of the static library.
```
Should `getRuntimeStaticLibraryPath` just do this? Are there places where we call it and need it to not do this?

troughton replied:
Windows is the only place we currently add the arch that I can see. I *think* on Darwin universal binaries are produced and copied to the lib/swift_static/<platform> folder; on Linux, I think the current assumption is that only one architecture will ever be present per platform, and so just using the copied libraries is fine (although probably not *technically* correct). 

On Windows, the files in that folder are (currently) just the static libraries for whichever arch was built last, so linking against that is definitely incorrect.

jrose-apple replied:
So I think that's @zbowling's #13108. I think that pushes the arch up into `getRuntimeStaticLibraryPath` for everyone, so this would become redundant. But I'm not sure what state that PR is in, since we didn't get back to Zac very quickly and he hasn't touched it but then @alblue poked it again.

troughton replied:
That looks right. Once that's merged we can remove the extra code here and just use `getRuntimeStaticLibraryPathWithArch` instead.

zbowling replied:
yeah, sorry. I've been busying in a bit of freeze with all the code I posted in all the PRs in December to take them and try and cut and stabilize a usable toolchain to ship in Fuchsia. Everything is getting to a good place with that toolchain so I should have more time to switch back and rebase all my PRs to get them ready to land upstream easier (and so I can get a new toolchain going with changes since the 4.1 branch).

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1493,6 +1534,128 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+                                                    + getTriple().getArchName()));
+
+    Arguments.push_back("-Xlinker");
+    Arguments.push_back("/NODEFAULTLIB:libcmt");
```
Can you leave a comment for why this is necessary?

troughton replied:
Done. I'm not actually certain it'll be necessary in the future (it may just be an artefact of how static libraries are currently produced for Windows), but for now adding that flag avoids the mentioned issues.

jrose-apple reviewed (commented):
Don't see anything that raises any flags for me. Yes, if Saleem's good I'm good. Thanks, Thomas.

compnerd commented:
@troughton - any chance you can rebase/squash the patches into a coherent patch?

troughton commented:
@compnerd I'll take a look – I should have some time to do that today. While you're around, can you quickly look at the `/NODEFAULTLIB:libcmt` flag and confirm that it _should_ be necessary? It was definitely needed last time I built, but that could be because of issues/flags elsewhere in the build that have since been resolved.

troughton committed:
Implement a Windows toolchain.

troughton commented:
@compnerd Provided this all looks okay to you, we should be ready to test and merge.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1510,6 +1542,136 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  }
+  if (!Linker.empty()) {
+    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + Linker));
+  }
```
The braces are unnecessary.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1510,6 +1542,136 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    //
+    // To avoid this, we specify that libcmt should not be linked against.
+    Arguments.push_back("-Xlinker");
+    Arguments.push_back("/NODEFAULTLIB:libcmt");
```
This seems wrong.  We should be building with `-MT` or `-MD` appropriately.

troughton replied:
I've removed this for now. It's currently necessary to build (at least in my case), but I think it's symptomatic of a bigger issue: how do we decide which library is appropriate? 

Currently, it's whatever the standard library is linked against, but I don't think that's right; rather, it should be determined based on the product we're building. Is the right solution to build the standard library with undefined symbols that are resolved once it's linked with the final Swift executable/library?

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1510,6 +1542,136 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  }
+
+  if (!context.OI.SDKPath.empty()) {
+    Arguments.push_back("--sysroot");
```
This feels like it won't work.  The toolchain for `x86_64-unknown-windows-msvc` does not support `--sysroot`.

troughton replied:
I've replaced `--sysroot` with `-I`.

compnerd reviewed (commented):
Please `clang-format` the diff

compnerd commented:
Pretty excited about this, I think that this will help with getting the IRGen tests passing, and I can look into fixing the rest of the Driver tests as well.

compnerd commented:
Let's go ahead and merge this and then we can improve it from there.

@swift-ci please test

troughton commented:
The build failure above appears to be unrelated, but I'd missed a test when I changed `--sysroot` to `-I`; the above commit was just fixing that.

jrose-apple commented:
@swift-ci Please test

troughton commented:
The `linker-autolink-extract` test failed on Linux due to the added Windows lines; I was actually testing something that isn't yet implemented, and missed it because I've been running the test suite on macOS. 

I've removed the changes to `linker-autolink-extract` for now.

compnerd commented:
`swift-autolink-extract` should not even exist on Windows.  That is a result of missing functionality in Linux.  In fact, I have pending working to remove that from Linux even.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test
